packages/astro/src/core/create-vite.ts

```diff
@@ -131,111 +129,31 @@ export async function createVite(
 	let result = commonConfig;
 	result = vite.mergeConfig(result, astroConfig.vite || {});
 	result = vite.mergeConfig(result, commandConfig);
-	sortPlugins(result);
+	if (result.plugins) {
+		sortPlugins(result.plugins);
+	}
 
 	return result;
 }
 
-function getPluginName(plugin: vite.PluginOption) {
-	if (plugin && typeof plugin === 'object' && !Array.isArray(plugin)) {
-		return plugin.name;
-	}
+function isVitePlugin(plugin: vite.PluginOption): plugin is vite.Plugin {
+	return Boolean(plugin?.hasOwnProperty('name'));
 }
 
-function sortPlugins(result: ViteConfigWithSSR) {
-	// HACK: move mdxPlugin to top because it needs to run before internal JSX plugin
-	const mdxPluginIndex =
-		result.plugins?.findIndex((plugin) => getPluginName(plugin) === '@mdx-js/rollup') ?? -1;
-	if (mdxPluginIndex === -1) return;
-	const jsxPluginIndex =
-		result.plugins?.findIndex((plugin) => getPluginName(plugin) === 'astro:jsx') ?? -1;
-	const mdxPlugin = result.plugins?.[mdxPluginIndex];
-	result.plugins?.splice(mdxPluginIndex, 1);
-	result.plugins?.splice(jsxPluginIndex, 0, mdxPlugin);
+function findPluginIndexByName(pluginOptions: vite.PluginOption[], name: string): number {
+	return pluginOptions.findIndex(function (pluginOption) {
+		// Use isVitePlugin to ignore nulls, booleans, promises, and arrays
+		// CAUTION: could be a problem if a plugin we're searching for becomes async!
+		return isVitePlugin(pluginOption) && pluginOption.name === name
+	})
 }
 
-// Scans `projectRoot` for third-party Astro packages that could export an `.astro` file
-// `.astro` files need to be built by Vite, so these should use `noExternal`
-async function getAstroPackages({ root }: AstroConfig): Promise<string[]> {
```

Comment:
Don't we still need this? Without this, every user will have to add any dependency with `.astro` files to `noExternal` manually. This was being used to make that process automatic.

Reply:
Right you are! Got trigger-happy with the refactors. Adding back 👍